packages/api/src/middleware/validation/checkAnimalOrBatch.js

```diff
@@ -13,81 +13,579 @@
+    checkIsArray(animalOrBatch[relationshipsKey], relationshipsKey);
+    await checkOtherUseRelationshipNotes(animalOrBatch[relationshipsKey]);
+  }
+};
+
+const checkAnimalOrigin = async (animalOrBatch, creating = true) => {
+  const { origin_id, brought_in_date } = animalOrBatch;
+  if (oneExists(['origin_id', 'brought_in_date'], animalOrBatch)) {
+    const broughtInOrigin = await AnimalOriginModel.query().where({ key: 'BROUGHT_IN' }).first();
+    // Overwrite date with null in db if editing origin_id
+    if (!creating && origin_id != broughtInOrigin.id) {
+      setFalsyValuesToNull(['brought_in_date'], animalOrBatch);
+    }
+
+    if (origin_id != broughtInOrigin.id && brought_in_date) {
+      throw customError('Brought in date must be used with brought in origin');
+    }
+  }
+};
+
+const checkAndAddCustomTypesOrBreeds = (
+  animalOrBatch,
+  newTypesSet,
+  newBreedsSet,
+  animalOrBatchRecord = undefined,
+) => {
+  const {
+    type_name,
+    breed_name,
+    custom_type_id,
+    default_type_id,
+    default_breed_id,
+    custom_breed_id,
+  } = animalOrBatch;
+  if (type_name) {
+    let defaultBreedId = default_breed_id;
+    let customBreedId = custom_breed_id;
+
+    if (!oneExists(['default_breed_id', 'custom_breed_id'], animalOrBatch) && animalOrBatchRecord) {
+      defaultBreedId = animalOrBatchRecord.default_breed_id;
+      customBreedId = animalOrBatchRecord.custom_breed_id;
+    }
+
+    if (defaultBreedId || customBreedId) {
+      throw customError('Cannot create a new type associated with an existing breed');
+    }
+    newTypesSet.add(type_name);
+  }
+
+  // newBreedsSet will be used to check if the combination of type + breed exists in DB.
+  // skip the process if the type is new (= type_name is passed)
+  if (!type_name && breed_name) {
+    let defaultTypeId = default_type_id;
+    let customTypeId = custom_type_id;
+
+    if (!oneExists(['default_type_id', 'custom_type_id'], animalOrBatch) && animalOrBatchRecord) {
+      defaultTypeId = animalOrBatchRecord.default_type_id;
+      customTypeId = animalOrBatchRecord.custom_type_id;
+    }
+
+    const breedDetails = customTypeId
+      ? `custom_type_id/${customTypeId}/${breed_name}`
+      : `default_type_id/${defaultTypeId}/${breed_name}`;
+
+    newBreedsSet.add(breedDetails);
+  }
+};
+
+const checkRemovalDataProvided = (animalOrBatch) => {
+  const { animal_removal_reason_id, removal_date } = animalOrBatch;
+  if (!animal_removal_reason_id || !removal_date) {
+    throw customError('Must send reason and date of removal');
+  }
+};
+
+const getRecordIfExists = async (animalOrBatch, animalOrBatchKey, farm_id) => {
+  return await AnimalOrBatchModel[animalOrBatchKey]
+    .query()
+    .findById(animalOrBatch.id)
+    .where({ farm_id })
+    .whereNotDeleted()
+    .withGraphFetched({
+      group_ids: true,
+      sex_detail: animalOrBatchKey === 'batch' ? true : false,
+      animal_batch_use_relationships: true,
```

Comment:
```suggestion
      ...(animalOrBatchKey === 'batch' && { sex_detail: true }),
      ...(animalOrBatchKey === 'animal'
        ? { animal_use_relationships: true }
        : { animal_batch_use_relationships: true }),
    });
```

@Duncan-Brain I'm getting errors here when I send an animal since these properties are not defined at all on that model -- I think `sex_detail` needs to be omitted, not set to false and the other property name is dependent on model.

Reply:
Yeah I started with batch. Now copy pasting over to animals and seeing what breaks. I saw the same errors on test failure

Reply:
Should be able to see that fix now if you needed it.

Reply:
I am just working on the "how to review this pr" guide but the structure should be solidified now.

@kathyavini For your earlier question about providing the exact same data to edit from frontend as pre-existing data it should be okay!

Here is a test I wrote that passes for animal batch --  you can add on to it and try if you want to test yourself -- or if you think I should add this in to my tests as valuable I can do that too:

```
 {
    testName: 'Edit everything and provide exact same data to patch with no errors',
    getRawRecordMismatch: (existingBatches) => {
      return {
        model: AnimalBatchUseRelationshipModel,
        where: { animal_batch_id: existingBatches[0].id },
        getMatchingBody: (existingBatches, records) => {
          return [
            {
              ...records[0],
            },
          ];
        },
      };
    },
    getPatchBody: (batch, existingBatches) => [
      {
        ...existingBatches[0]
      },
    ],
    getPostBody: () => [
      {
        name: 'edit test 1',
        default_type_id: defaultTypeId,
        animal_batch_use_relationships: [{ use_id: animalUse2.id }],
        sire: 'Unchanged',
        count: 4,
        sex_detail: [
          {
            sex_id: animalSex1.id,
            count: 2,
          },
          {
            sex_id: animalSex2.id,
            count: 2,
          },
        ],
      },
    ],
  },
```